File: web/src/lib/scoring.ts
import type { FeatureVector, MissingElement, Severity, WeightedScore } from "./types";

export const BIAS_NOTICE =
  "Historical crime datasets may reflect systemic bias in policing and prosecution. " +
  "SDS models must be audited for fairness before operational use and must never " +
  "consider protected characteristics (race, religion, neighbourhood proxies, etc.). " +
  "Predictions are supportive analytics only — not legal judgments.";

const EVIDENCE_WEIGHTS: Record<string, [number, string]> = {
  forensicDna: [0.95, "DNA / genetic evidence"],
  forensicFingerprint: [0.85, "Fingerprint evidence"],
  confessionRecorded: [0.8, "Recorded confession (lawfully obtained)"],
  cctvPresent: [0.75, "CCTV / video corroboration"],
  weaponRecovered: [0.7, "Physical weapon link"],
  credibleEyewitnesses: [0.65, "Independent eyewitnesses"],
  identificationQuality: [0.6, "Identification procedure quality"],
  expertForensic: [0.6, "Expert / laboratory report"],
  chainOfCustody: [0.55, "Exhibit continuity"],
  statementQuality: [0.5, "Witness statement quality"],
  disclosureComplete: [0.45, "Disclosure completeness"],
  timeToArrestScore: [0.35, "Investigation timeliness"],
  hearsayRisk: [-0.5, "Hearsay exposure (penalty)"],
};

function normalize(key: keyof FeatureVector, value: number): number {
  if (key === "credibleEyewitnesses") return Math.min(1, value / 3);
  return Math.max(0, Math.min(1, value));
}

export function computeWeightedScores(vector: FeatureVector): WeightedScore[] {
  const scores: WeightedScore[] = [];

  for (const [key, [weight, label]] of Object.entries(EVIDENCE_WEIGHTS)) {
    const raw = vector[key as keyof FeatureVector];
    const normalized = key === "hearsayRisk" ? raw : normalize(key as keyof FeatureVector, raw);
    const contribution = weight * normalized;
    scores.push({
      category: label,
      weight,
      score: normalized,
      weightedContribution: contribution,
      note:
        key === "hearsayRisk"
          ? "Penalty applied when hearsay language appears."
          : `Weighted contribution: ${contribution.toFixed(2)}`,
    });
  }

  return scores.sort((a, b) => Math.abs(b.weightedContribution) - Math.abs(a.weightedContribution));
}

export function compositeScoreFromWeights(scores: WeightedScore[]): number {
  const totalWeight = scores.filter((s) => s.weight > 0).reduce((sum, s) => sum + s.weight, 0);
  if (!totalWeight) return 0;
  const raw = scores.reduce((sum, s) => sum + s.weightedContribution, 0);
  return Math.max(0, Math.min(100, Math.round((raw / totalWeight) * 100)));
}

export function detectMissingElements(vector: FeatureVector): MissingElement[] {
  const missing: MissingElement[] = [];
  const forensicOk =
    vector.forensicDna >= 1 || vector.forensicFingerprint >= 1 || vector.expertForensic >= 1;
  const witnessOk = vector.credibleEyewitnesses >= 1 || vector.identificationQuality >= 0.5;

  const checks: { ok: boolean; element: string; severity: Severity; detail: string }[] = [
    {
      ok: forensicOk,
      element: "Forensic link (DNA, fingerprint, or expert report)",
      severity: "high",
      detail: "No clear forensic or expert laboratory link detected in the file text.",
    },
    {
      ok: witnessOk,
      element: "Independent eyewitness or strong identification",
      severity: "high",
      detail: "No credible eyewitness or robust identification procedure referenced.",
    },
    {
      ok: vector.chainOfCustody >= 0.5,
      element: "Exhibit chain of custody",
      severity: "medium",
      detail: "Continuity of physical exhibits is not documented in the uploaded text.",
    },
    {
      ok: vector.statementQuality >= 0.5,
      element: "Lawful recorded statement",
      severity: "medium",
      detail: "No indication of a properly recorded and cautioned statement.",
    },
    {
      ok: vector.disclosureComplete >= 1,
      element: "Disclosure schedule",
      severity: "medium",
      detail: "Disclosure / unused-material review not evidenced in the bundle.",
    },
  ];

  for (const item of checks) {
    if (!item.ok) {
      missing.push({ element: item.element, severity: item.severity, detail: item.detail });
    }
  }
  return missing;
}

File: sds/scoring.py
from __future__ import annotations

from .features import FeatureVector
from .models import MissingElement, Severity, WeightedScore

# Evidence weights — high for forensic/DNA, lower for circumstantial signals.
EVIDENCE_WEIGHTS: dict[str, tuple[float, str]] = {
    "forensic_dna": (0.95, "DNA / genetic evidence"),
    "forensic_fingerprint": (0.85, "Fingerprint evidence"),
    "confession_recorded": (0.80, "Recorded confession (lawfully obtained)"),
    "cctv_present": (0.75, "CCTV / video corroboration"),
    "weapon_recovered": (0.70, "Physical weapon link"),
    "credible_eyewitnesses": (0.65, "Independent eyewitnesses"),
    "identification_quality": (0.60, "Identification procedure quality"),
    "expert_forensic": (0.60, "Expert / laboratory report"),
    "chain_of_custody": (0.55, "Exhibit continuity"),
    "statement_quality": (0.50, "Statement process quality"),
    "disclosure_complete": (0.45, "Disclosure completeness"),
    "time_to_arrest_score": (0.35, "Investigation timeliness"),
    "hearsay_risk": (-0.50, "Hearsay exposure (penalty)"),
}

CONVICTION_REQUIREMENTS: tuple[tuple[str, str, Severity, str], ...] = (
    (
        "forensic_dna",
        "Forensic link (DNA, fingerprint, or expert report)",
        Severity.HIGH,
        "No clear forensic or expert laboratory link detected in the file text.",
    ),
    (
        "credible_eyewitnesses",
        "Independent eyewitness or strong identification",
        Severity.HIGH,
        "No credible eyewitness or robust identification procedure referenced.",
    ),
    (
        "chain_of_custody",
        "Exhibit chain of custody",
        Severity.MEDIUM,
        "Continuity of physical exhibits is not documented in the uploaded text.",
    ),
    (
        "statement_quality",
        "Lawful recorded statement",
        Severity.MEDIUM,
        "No indication of a properly recorded and cautioned statement.",
    ),
    (
        "disclosure_complete",
        "Disclosure schedule",
        Severity.MEDIUM,
        "Disclosure / unused-material review not evidenced in the bundle.",
    ),
)


def compute_weighted_scores(vector: FeatureVector) -> list[WeightedScore]:
    features = vector.as_dict()
    scores: list[WeightedScore] = []

    for key, (weight, label) in EVIDENCE_WEIGHTS.items():
        raw = features[key]
        if key == "credible_eyewitnesses":
            normalized = min(1.0, raw / 3.0)
        elif key == "hearsay_risk":
            normalized = raw
            contribution = weight * normalized
            scores.append(
                WeightedScore(
                    category=label,
                    weight=weight,
                    score=normalized,
                    weighted_contribution=contribution,
                    note="Penalty applied when hearsay language appears.",
                )
            )
            continue
        else:
            normalized = max(0.0, min(1.0, raw))

        contribution = weight * normalized
        scores.append(
            WeightedScore(
                category=label,
                weight=weight,
                score=normalized,
                weighted_contribution=contribution,
                note=f"Weighted contribution: {contribution:.2f}",
            )
        )

    scores.sort(key=lambda s: abs(s.weighted_contribution), reverse=True)
    return scores


def composite_score_from_weights(scores: list[WeightedScore]) -> int:
    total_weight = sum(abs(s.weight) for s in scores if s.weight > 0)
    if total_weight == 0:
        return 0
    raw = sum(s.weighted_contribution for s in scores)
    max_possible = total_weight
    pct = (raw / max_possible) * 100
    return max(0, min(100, round(pct)))


def detect_missing_elements(vector: FeatureVector) -> list[MissingElement]:
    features = vector.as_dict()
    missing: list[MissingElement] = []

    forensic_ok = (
        features["forensic_dna"] >= 1
        or features["forensic_fingerprint"] >= 1
        or features["expert_forensic"] >= 1
    )
    witness_ok = features["credible_eyewitnesses"] >= 1 or features["identification_quality"] >= 0.5

    checks = {
        "forensic_dna": forensic_ok,
        "credible_eyewitnesses": witness_ok,
        "chain_of_custody": features["chain_of_custody"] >= 0.5,
        "statement_quality": features["statement_quality"] >= 0.5,
        "disclosure_complete": features["disclosure_complete"] >= 1,
    }

    for key, label, severity, detail in CONVICTION_REQUIREMENTS:
        if not checks.get(key, False):
            missing.append(MissingElement(element=label, severity=severity, detail=detail))

    return missing
